server/modlog.ts

```diff
@@ -36,284 +32,279 @@ const PUNISHMENTS = [
-			}
-		})();
-	}
-	return Config.ripgrepmodlog;
+export interface ModlogEntry {
+	action: string;
+	roomID?: string;
+	visualRoomID?: string;
+	userid?: ID;
+	autoconfirmedID?: ID;
+	alts?: ID[];
+	ip?: string;
+	isGlobal?: boolean;
+	loggedBy?: ID;
+	note?: string;
+	/** Milliseconds since the epoch */
+	time?: number;
 }
 
 export class Modlog {
-	readonly logPath: string;
-	/**
-	 * If a stream is undefined, that means it has not yet been initialized.
-	 * If a stream is truthy, it is open and ready to be written to.
-	 * If a stream is null, it has been destroyed/disabled.
-	 */
-	sharedStreams: Map<ID, Streams.WriteStream | null> = new Map();
-	streams: Map<ModlogID, Streams.WriteStream | null> = new Map();
+	readonly database: Database.Database;
+	readonly usePM: boolean;
+	readonly modlogInsertionQuery: Database.Statement<ModlogEntry>;
+	readonly altsInsertionQuery: Database.Statement<[number, string]>;
+	readonly renameQuery: Database.Statement<[string, string]>;
+	readonly globalPunishmentsSearchQuery: Database.Statement<[string, string, string, number, ...string[]]>;
+	readonly insertionTransaction: Database.Transaction;
+
+	constructor(path: string, noProcessManager = false) {
+		this.database = new Database(path);
+		this.database.exec("PRAGMA foreign_keys = ON;");
+		this.database.exec(FS(MODLOG_SCHEMA_PATH).readIfExistsSync()); // Set up tables, etc.
+		this.database.function('regex', {deterministic: true}, (regexString, toMatch) => {
+			return Number(RegExp(regexString).test(toMatch));
+		});
 
-	constructor(path: string) {
-		this.logPath = path;
+		this.modlogInsertionQuery = this.database.prepare(
+			`INSERT INTO modlog (timestamp, roomid, visual_roomid, action, userid, autoconfirmed_userid, ip, action_taker_userid, note)` +
+			` VALUES ($time, $roomID, $visualRoomID, $action, $userid, $autoconfirmedID, $ip, $loggedBy, $note)`
+		);
+		this.altsInsertionQuery = this.database.prepare(`INSERT INTO alts (modlog_id, userid) VALUES (?, ?)`);
+		this.renameQuery = this.database.prepare(`UPDATE modlog SET roomid = ? WHERE roomid = ?`);
+		this.globalPunishmentsSearchQuery = this.database.prepare(
+			`SELECT * FROM modlog WHERE (roomid = 'global' OR roomid LIKE 'global-%') ` +
+			`AND (userid = ? OR autoconfirmed_userid = ? OR EXISTS(SELECT * FROM alts WHERE alts.modlog_id = modlog.modlog_id AND userid = ?)) ` +
+			`AND timestamp > ?` +
+			`AND action IN (${this.formatArray(GLOBAL_PUNISHMENTS, [])}) `
+		);
+
+		this.insertionTransaction = this.database.transaction((entry: {
+			action: string,
+			roomID: string,
+			visualRoomID: string,
+			userid: ID,
+			autoconfirmedID: ID,
+			ip: string,
+			loggedBy: ID,
+			note: string,
+			time: number,
+			alts: ID[],
+		}) => {
+			const result = this.modlogInsertionQuery.run(entry);
+			for (const alt of entry.alts || []) {
+				this.altsInsertionQuery.run(result.lastInsertRowid as number, alt);
+			}
+		});
+		this.usePM = !noProcessManager && !Config.nofswriting;
```

Comment:
I don't know what `nofswriting` controls here, if it means we don't write to a DB then I don't know why we are opening a database at all or creating queries

Reply:
`nofswriting` indicates that we are in a test environment; we don't want to use a multi-process modlog setup in a test environment but we can still query an in-memory database.